pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,120 @@
+from datetime import timedelta
+import numbers
+from typing import (
+    Any,
+    Generic,
+    TypeVar,
+    overload,
+)
+
+import numpy as np
+import numpy.typing as npt
+
+from pandas._libs import (
+    Timedelta,
+    Timestamp,
+)
+from pandas._typing import IntervalBound
+
+_OrderableMixinT = TypeVar(
+    "_OrderableMixinT", int, float, Timestamp, Timedelta, npt.NDArray[np.generic]
+)
+_OrderableT = TypeVar("_OrderableT", int, float, Timestamp, Timedelta)
+
+# note: mypy doesn't support overloading properties
+# based on github.com/microsoft/python-type-stubs/pull/167
+class _LengthProperty:
+    @overload
+    def __get__(self, instance: IntervalMixin[Timestamp], owner: Any) -> Timedelta: ...
+    @overload
+    def __get__(
+        self, instance: IntervalMixin[_OrderableMixinT], owner: Any
+    ) -> _OrderableMixinT: ...
+
+class IntervalMixin(Generic[_OrderableMixinT]):
+    @property
+    def closed_left(self) -> bool: ...
+    @property
+    def closed_right(self) -> bool: ...
+    @property
+    def open_left(self) -> bool: ...
+    @property
+    def open_right(self) -> bool: ...
+    @property
+    def mid(self) -> _OrderableT: ...
+    length: _LengthProperty
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...) -> None: ...
+
+class Interval(IntervalMixin[_OrderableT]):
+    def __init__(
+        self,
```

Comment:
probably need to use `__new__()` here to properly classify the interval types.

Reply:
Are cython's `__init__` exposed as `__new__` (there is no `__new__` in `interval.pyx`)? Your example seems to work fine with the current code:

```py
import pandas as pd

time: pd.Timestamp = pd.Timestamp(0)
i1 = pd.Interval(time, time, closed="both")
reveal_type(i1)
reveal_type(i1.left)
reveal_type(i1.length)

i2 = pd.Interval(10, 20)
reveal_type(i2)
reveal_type(i2.left)
reveal_type(i2.length)

i3 = pd.Interval(13.2, 19.5)
reveal_type(i3)
reveal_type(i3.left)
reveal_type(i3.length)
```

Reply:
Somehow I was misled thinking I had to use `__new__` to get the types right, but I just switched the MS PR to use `__init__()` and it works.